torch/cuda/streams.py

```diff
@@ -119,90 +102,100 @@ def __repr__(self):
                 .format(self.device, self.cuda_stream))
 
 
-class EventHandle(ctypes.Structure):
-    IPC_HANDLE_SIZE = 64
-    _fields_ = [('reserved', ctypes.c_char * IPC_HANDLE_SIZE)]
+class Event(torch._C._CudaEventBase):
+    r"""Wrapper around a CUDA event.
 
+    CUDA events are synchronization markers that can be used to monitor the
+    device's progress, to accurately measure timing, and to synchronize CUDA
+    streams.
 
-class Event(object):
-    r"""Wrapper around CUDA event.
+    The underlying CUDA events are lazily initialized when the event is first
+    recorded or exported to another process. After creation, only streams on the
+    same device may record the event. However, streams on any device can wait on
+    the event.
 
     Arguments:
-        enable_timing (bool): indicates if the event should measure time
+        device(torch.device or int, optional): a device on which to allocate
+            the event. If :attr:`device` is ``None`` (default) or a negative
+            integer, this will use the current device.
+        enable_timing (bool, optional): indicates if the event should measure time
             (default: ``False``)
-        blocking (bool): if ``True``, :meth:`wait` will be blocking (default: ``False``)
+        blocking (bool, optional): if ``True``, :meth:`wait` will be blocking (default: ``False``)
         interprocess (bool): if ``True``, the event can be shared between processes
             (default: ``False``)
+
+       .. _CUDA documentation:
+       https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EVENT.html
     """
 
-    DEFAULT = 0x0
-    BLOCKING_SYNC = 0x1
-    DISABLE_TIMING = 0x2
-    INTERPROCESS = 0x4
-
-    def __init__(self, enable_timing=False, blocking=False, interprocess=False,
-                 _handle=None):
-        flags = Event.DEFAULT
-        if not enable_timing:
-            flags |= Event.DISABLE_TIMING
-        if blocking:
-            flags |= Event.BLOCKING_SYNC
-        if interprocess:
-            flags |= Event.INTERPROCESS
-
-        ptr = ctypes.c_void_p()
-        self._cudart = cudart()
-        if _handle:
-            check_error(self._cudart.cudaIpcOpenEventHandle(ctypes.byref(ptr), _handle))
-        else:
-            check_error(self._cudart.cudaEventCreateWithFlags(ctypes.byref(ptr), ctypes.c_uint(flags)))
-        self._as_parameter_ = ptr
-
-    def __del__(self):
-        if hasattr(self, '_as_parameter_'):
-            check_error(self._cudart.cudaEventDestroy(self._as_parameter_))
-            del self._as_parameter_
+    def __new__(cls, device=None,
```

Comment:
I don't think it makes sense to take the `device` as an argument when the event is lazily constructed, so it may just end up getting ignored. It makes sense if your importing an even through a ipc_handle. Since this constructor is getting complicated -- not all arguments make sense together -- I think it's worth splitting out the construction of an event from an IPC handle into a separate factory function:

Something like

```
def from_ipc_handle(device, ipc_handle)
```

(Probably implemented in C++)

Reply:
(This will require changing `rebuild_event` and `reduce_event` in torch/multiprocessing/reductions.py)